group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -2201,75 +2222,85 @@ private CoordinatorResult<StreamsGroupHeartbeatResult, CoordinatorRecord> stream
             records
         );
 
-        if (group.topology() != null && updatedMember.topologyEpoch() < group.topology().topologyEpoch()) {
-            returnedStatus.add(
-                new Status()
-                    .setStatusCode(StreamsGroupHeartbeatResponse.Status.STALE_TOPOLOGY.code())
-                    .setStatusDetail(
-                        String.format(
-                            "The member's topology epoch %d is behind the group's topology epoch %d.",
-                            updatedMember.topologyEpoch(),
-                            group.topology().topologyEpoch()
-                        )
-                    )
-            );
-        }
-
-        // 2. Initialize/Update the group topology.
-        if (clientTopology != null) {
+        // 2. Initialize/Update the group topology. If the member is new or has changed, a StreamsGroupTopologyValue
+        // record is written to the __consumer_offsets partition to persist the change. The group epoch is bumped if the
+        // topology has changed.
+        StreamsTopology updatedTopology;
+        boolean reconfigureTopology = false;
+        if (topology != null) {
+            throwIfInvalidTopology(topology);
 
-            StreamsGroupTopologyValue recordValue = convertToStreamsGroupTopologyRecord(clientTopology);
+            StreamsGroupTopologyValue recordValue = convertToStreamsGroupTopologyRecord(topology);
 
             final Map<String, StreamsGroupTopologyValue.Subtopology> subtopologyMap = recordValue.subtopologies().stream()
                 .collect(Collectors.toMap(StreamsGroupTopologyValue.Subtopology::subtopologyId, x -> x));
 
-            topology = new StreamsTopology(clientTopology.epoch(), subtopologyMap);
+            updatedTopology = new StreamsTopology(topology.epoch(), subtopologyMap);
 
             if (group.topology() == null) {
-                log.info("[GroupId {}][MemberId {}] Member initialized the topology with epoch {}", groupId, memberId, clientTopology.epoch());
+                log.info("[GroupId {}] Member {} initialized the topology with epoch {}", groupId, memberId, topology.epoch());
```

Comment:
Yeah that was just a rebase mistake. 

Although I do wonder if we really got to a "uniform log messages", because now the log messages between the group types are inconsistent....

Fixed